### PR TITLE
Faster twin renderer

### PR DESCRIPTION
**Pull request by carolhmj**
This PR is a rearchitecture of the HTMLTwinRenderer, aiming to improve its speed. 

Some PGs I tested with:
The ones in: https://doc.babylonjs.com/toolsAndResources/accessibility/screenReaders
Stress test: #C9GZTF#43
AddAllControls: #C9GZTF#49
Weapons demo with accessibility: #I6V1ST#240

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14441/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14441/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14441/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14441/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14441/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14441/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14441/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14441/merge#BCU1XR#0

File: packages/tools/accessibility/src/HtmlTwin/htmlTwinItemAdapter.tsx [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    const sceneContext = useContext(SceneContext);
+    useEffect(() => {
+        const observable = (node as any).onDisposeObservable;
+        const observer = observable.add(() => {
+            sceneContext.updateScene();
+        });
+        return () => {
+            observable.remove(observer);
+        };
+    }, [node]);
```
Comment (RaananW) [inline packages/tools/accessibility/src/HtmlTwin/htmlTwinItemAdapter.tsx:52]: can (and should) be unified. it is the same conditional useEffect

File: packages/tools/accessibility/src/HtmlTwin/htmlTwinItemAdapter.tsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        useEffect(() => {
+            const observable = node.onIsVisibleChangedObservable;
+            const observer = observable.add(() => {
+                setIsVisibleState(isVisible(props.node));
+            });
+            return () => {
+                observable.remove(observer);
+            };
+        }, [node]);
+    }
```
Comment (RaananW) [inline packages/tools/accessibility/src/HtmlTwin/htmlTwinItemAdapter.tsx:100]: useEffect should not be conditional. You can check for the instanceof inside the useEffect callback instead.

File: packages/tools/accessibility/src/HtmlTwin/htmlTwinSceneTree.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            textures.push(texture);
+        }
+    }
+    return textures;
+}
+
+export function HTMLTwinSceneTree(props: { scene: Scene; options: IHTMLTwinRendererOptions }) {
+    const { scene, options } = props;
+
+    const [, setMeshIds] = useState(new Set<number>());
```
Comment (RaananW) [inline packages/tools/accessibility/src/HtmlTwin/htmlTwinSceneTree.tsx:30]: trying to understand - why only a setter is needed? is the value not used anywhere?

**Review by RaananW — COMMENTED** [review]
Let's discuss this offline :-)

File: packages/tools/accessibility/src/HtmlTwin/htmlTwinItemAdapter.tsx [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    useEffect(() => {
+        setDescription(twinItem?.getDescription(options));
+    }, [twinItem]);
+
+    const [children, setChildren] = useState(getDirectChildrenOf(props.node));
+    if (node instanceof Container) {
+        useEffect(() => {
+            const observable = node.onControlAddedObservable;
+            const observer = observable.add(() => {
+                setChildren([...getDirectChildrenOf(props.node)]);
```
Comment (Popov72) [inline packages/tools/accessibility/src/HtmlTwin/htmlTwinItemAdapter.tsx:74]: Do you have to create a new array when calling `setChildren`? If yes, I think that doing `getDirectChildrenOf(props.node).slice()` is better for performance.

File: packages/tools/accessibility/src/HtmlTwin/htmlTwinSceneContext.tsx [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { createContext } from "react";
+
+export interface ISceneContext {
```
Comment (Popov72) [inline packages/tools/accessibility/src/HtmlTwin/htmlTwinSceneContext.tsx:3]: nit: missing doc

**Review by Popov72 — COMMENTED** [review]
Mostly missing doc!

Generally speaking, there are a lot of checks with "instanceof", which is slow in javascript. If they're not in the hot path, then that's fine, otherwise I think you should try replacing them with something else (when possible), because as I understand it, this PR is intended to improve overall performance?

File: packages/tools/accessibility/src/HtmlTwin/htmlTwinItem.tsx [hunk 6]
```diff
@@ -1,9 +1,46 @@
...
+            }
+        }
+    }
+    return undefined;
+}
+
+export function getDirectChildrenOf(item: AccessibilityEntity): AccessibilityEntity[] {
+    if (item instanceof Node) {
+        return item.getDescendants(true);
+    } else if (item instanceof Container && !(item instanceof Button)) {
```
[the remark this answers is not in the document]
Comment (carolhmj) [reply packages/tools/accessibility/src/HtmlTwin/htmlTwinItem.tsx:29]: All Containers work the same (except for the button), so this comparison is a bit simpler

File: packages/tools/accessibility/src/HtmlTwin/htmlTwinItemAdapter.tsx [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        return null;
+    }
+    const [twinItem, setTwinItem] = useState<HTMLTwinItem>(getTwinItemFromNode(node, scene));
+    useEffect(() => {
+        setTwinItem(getTwinItemFromNode(node, scene));
+    }, [node]);
+
+    const [isVisibleState, setIsVisibleState] = useState(isVisible(props.node));
+    useEffect(() => {
+        const observable = (node as any).onEnabledStateChangedObservable;
```
[the remark this answers is not in the document]
Comment (carolhmj) [reply packages/tools/accessibility/src/HtmlTwin/htmlTwinItemAdapter.tsx:34]: It exists on Node too :D

**Review by RaananW — APPROVED** [review]

**Review by Popov72 — APPROVED** [review]

**Pull request merged**